How-To: .NET Interactive + .NET for Apache Spark (#20926)
dotnet/docs: docs/spark/how-to-guides/dotnet-spark-jupyter-notebooks.md

---
title: Use Jupyter notebooks
titleSuffix: .NET for Apache Spark
description: Use .NET for Apache Spark in interactive environments like Jupyter Notebook, Jupyter Lab, or Visual Studio Code (VS Code)
ms.author: luquinta
author: luisquintanilla
ms.date: 10/09/2020
ms.topic: conceptual
ms.custom: mvc, how-to
---

# Use .NET for Apache Spark in Jupyter notebooks

In this article, you learn how to run .NET for Apache Spark jobs interactively in Jupyter Notebook and Visual Studio Code (VS Code) with .NET Interactive.

## About Jupyter

[Jupyter](https://jupyter.org/) is an open-source, cross-platform computing environment that provides a way for users to prototype and develop applications interactively. You can interact with Jupyter through a wide variety of interfaces such as Jupyter Notebook, Jupyter Lab, and VS Code.

In the context of .NET, [.NET Interactive](https://github.com/dotnet/interactive), a .NET Core global tool, provides a kernel for writing .NET code (C#/F#) in interactive computing environments such as Jupyter Notebook.

## Prerequisites

- [.NET Core 3.1 SDK](https://docs.microsoft.com/dotnet/core/install/)

> - Apache Spark

Reviewer: Can we add a link to Apache Spark official website or something here?

Author: done

> - Apache Spark .NET Worker

Reviewer: Let's also add a link here to the [Spark .NET github doc](https://github.com/dotnet/spark/blob/master/deployment/README.md#preparing-worker-dependencies) for installing worker.

Author: done

See the [getting started tutorial](../tutorials/get-started.md) for more information on setting up your .NET for Apache Spark environment.

## Prepare environment

To work with Jupyter Notebooks, you'll need two things.

1. Install [.NET Interactive global .NET tool](https://github.com/dotnet/interactive/blob/main/docs/NotebooksLocalExperience.md)
1. Download `Microsoft.Spark` NuGet package.
    1. Navigate to the [Microsoft.Spark](https://www.nuget.org/packages/Microsoft.Spark/) NuGet package page.

        > [!IMPORTANT]
        > By default, the latest version of the package is downloaded. **Make sure that the version you download is the same as your Apache Spark .NET Worker.**

    1. In the **Info** pane, select **Download package**. Doing so downloads the latest version of the package to a file with a name similar to  *microsoft.spark.[PACKAGE-VERSION].nupkg*.
    1. Unzip the recently downloaded package. The unzipped directory should contain a subdirectory called *jars*. Take note of the path since it's used at a later time.

## Start .NET for Apache Spark

> Start .NET for Apache Spark in debug mode. This will will set up a process that will wait for connections from a [SparkSession](xref:Microsoft.Spark.Sql.SparkSession). Make sure to provide the path to the jar for the respective version of Apache Spark you're using.
> 

Reviewer: ```suggestion
Start .NET for Apache Spark in debug mode. This will will set up a process that will wait for connections from a [SparkSession](xref:Microsoft.Spark.Sql.SparkSession). Make sure to provide the path to the `microsoft-spark-<version>.jar` for the respective version of .NET for Apache Spark you're using.

```

Author: done

```bash
spark-submit \
  --class org.apache.spark.deploy.dotnet.DotnetRunner \
  --master local \
  <path-to-microsoft-spark-jar> \
  debug
```

## Create notebook

You can use different interfaces to interact with Jupyter. For a browser-based interface, use Jupyter Notebooks or Jupyter Lab. For a local editor experience, use VS Code.

### Jupyter Notebooks & Jupyter lab

1. In another command prompt, start Jupyter Notebook or Jupyter Lab use one of the commands below:

    **Jupyter Notebook**

    ```bash
    jupyter notebook
    ```

    **Jupyter Lab**

    ```bash
    jupyter lab
    ```

1. Create a new .NET Interactive notebook.

> 1. When the notebook opens, write and execute your code in the notebook cells.

Reviewer: Can we give a small example of how to do these steps and what a simple notebook should look like?

Author: Done

### Visual Studio Code (preview)

> [!IMPORTANT]
> To use Jupyter Notebooks in VS Code, you have to install:
>
>- [VS Code Insiders](https://code.visualstudio.com/insiders/)
>- [.NET Interactive Notebooks extension](https://marketplace.visualstudio.com/items?itemName=ms-dotnettools.dotnet-interactive-vscode)

> 1. Open VS Code

Reviewer: ```suggestion
1. Open VS Code.
```

Author: Done

> 1. Open the command palette **View > Command Palette**

Reviewer: ```suggestion
1. Open the command palette **View > Command Palette**.
```

Author: Done

    The command palette appears. Enter the following command to create a new .NET Interactive notebook:

    ```text
    .NET Interactive: Create new blank notebook
    ```

    Alternatively, if you want to open an existing .NET Interactive notebook with the *.ipynb* extension, use the following command:

    ```text
    .NET Interactive: Open notebook
    ```

> 1. When the notebook opens, write and execute your code in the notebook cells.

Reviewer: Ditto for here. I think it would be good to show through examples what this would look like.

Author: Done

## Next Steps

> - [Get started with .NET for Apache Spark](../tutorials/get-started.md)

Reviewer: Why not provide the link to the official website here instead -> [https://docs.microsoft.com/en-us/dotnet/spark/tutorials/get-started?tabs=windows](https://docs.microsoft.com/en-us/dotnet/spark/tutorials/get-started?tabs=windows)?

Author: Yup. This is where this link points to.

Author: Done

> - [Predict sentiment using .NET for Apache Spark and ML.NET](../tutorials/ml-sentiment-analysis.md)

Reviewer: Ditto, let's provide the official page link here as well.

Author: Similar to get started link comment, this should point to the tutorial

Author: Done

### Jupyter Notebooks & Jupyter lab

1. In another command prompt, start Jupyter Notebook or Jupyter Lab use one of the commands below:

    **Jupyter Notebook**

    ```bash
    jupyter notebook
    ```

    **Jupyter Lab**

    ```bash
    jupyter lab
    ```

    These commands launch a browser window with the Jupyter Notebook or Jupyter Lab interface.

1. In the browser, create a new notebook.

    **Jupyter Notebook**

    Select **New > .NET (C#)** or **New > .NET (F#)**

    **Jupyter Lab**

    In the Launcher window, select **.NET (C#)** or **.NET (F#)**

### Visual Studio Code (preview)

> [!IMPORTANT]
> To use Jupyter Notebooks in VS Code, you have to install:
>
>- [VS Code Insiders](https://code.visualstudio.com/insiders/)
>- [.NET Interactive Notebooks extension](https://marketplace.visualstudio.com/items?itemName=ms-dotnettools.dotnet-interactive-vscode)

1. Open VS Code.
1. Open the command palette **View > Command Palette**.

    The command palette appears. Enter the following command to create a new .NET Interactive notebook:

    ```text

>     .NET Interactive: Create new blank notebook

Author: You do, but it's automatically populated for you when you open the prompt so you only have to enter the text.

>     .NET Interactive: Create new blank notebook

Reviewer 2: oh, mine isn't configured that way - it defaults to files

    ```

    Alternatively, if you want to open an existing .NET Interactive notebook with the *.ipynb* extension, use the following command:

    ```text
    .NET Interactive: Open notebook
    ```

## Initialize Spark Session

> When the notebook opens
> 
> 1. Install the `Microsoft.Spark` NuGet package. Make sure the version you install is the same as the .NET Worker.

Reviewer 2: ```suggestion
1. When the notebook opens, install the `Microsoft.Spark` NuGet package. Make sure the version you install is the same as the .NET Worker.
```

    ```text
    #r "nuget:Microsoft.Spark, 0.12.1"
    ```

> 1. Add the following using statement

Reviewer 2: ```suggestion
1. Add the following using statement to the notebook.
```

    ```csharp
    using Microsoft.Spark.Sql;
    ```

> 1. Initialize your [SparkSession](xref:Microsoft.Spark.Sql.SparkSession)

Reviewer 2: ```suggestion
1. Initialize your [SparkSession](xref:Microsoft.Spark.Sql.SparkSession).
```

    ```csharp
    var sparkSession =
    SparkSession
        .Builder()
        .AppName("dotnet-interactive-spark")
        .GetOrCreate();
    ```

> The notebook should look similar to the one below. In this case, it's in VS Code, but Jupyter Notebook and Jupyter Lab should look about the same.

Reviewer 2: ```suggestion
The notebook should look similar to the one in the following image. This example uses VS Code, but Jupyter Notebook and Jupyter Lab should look about the same.
```
